PLEP-0007.rst

```diff
@@ -0,0 +1,847 @@
+  - Enable a simple simulation to be set up in ten to twenty lines of
+    code.
+
+  - Provide narrative documentation that progresses from a simulation
+    set up using the highest level of abstraction to lower levels of
+    abstraction.
+
+* Each numerical method should be programmed to a well-defined
+  interface.  PIC and fluid simulations should have identical
+  interfaces at the highest level of abstraction.
+  
+Functionality requirements
+--------------------------
+
+* Create a standardized representation for the problem setup so that
+  the same problem setup can be used interchangeably for different
+  physical models and numerical methods.
+
+.. Create a class or object that represents the dependent variables so
+   that it can be used when creating the other classes when needed?
+  
+  - Include distinct representations for the domain, ICs, and BCs.
+
+    - Check that different domain, IC, and BC representations are
+      consistent with each other.
+
+    - Provide clear, well-defined interfaces for each of these
+      representations.
+
+  - Allow distinct representation of volumetric source terms that are
+    needed to drive turbulence.
+
+  - Implement checks that the domain, ICs, and BCs are consistent with
+    each other.
+
+    - Use exception handling to provide useful error messages that
+      help users quickly pinpoint problems.
+
+  - Issue a warning when the initial conditions are far out of
+    equilibrium.
+
+  - Allow users to select pre-defined initial conditions that are
+    commonly used.
+
+* Incorporate grid generation capabilities.
+
+  - Implement general tools to create non-uniform logically
+    rectangular grids.
+
+  - Implement or use tools to create a finite element grid for
+    experimental geometries.
+
+* Prioritize flexible numerical methods that can solve almost
+  arbitrary systems of equations written in conservative form.
+
+  - Allow users to select pre-defined systems of equations.
+
+  - Allow users to provide equations as strings.
+  
+    - Parse each string to extract the flux and source terms for each
+      equation.
+
+    - Create functions for each of the fluxes and sources that can be
+      compiled at runtime (with automatic differentiation to calculate
+      Jacobians, when needed).
+
+  - Implement finite difference capabilities.
+
+  - Implement finite volume capabilities.
+
+  - Implement finite element/spectral element capabilities.
+
+  - Provide shock capturing algorithms.
+
+* Implement auxiliary functionality, including:
+
+  - Grad-Shafranov solver
+
+  - Synthetic diagnostics
+
+  - Magnetic topology analysis tools
+
+  - Turbulence analysis tools
+
+* Implement particle-in-cell simulation capabilities.
+
+  - Define abstract interfaces in the implementation to allow
+    different particle-pushers and other simulation components to be
+    used interchangeably.
+
+  - Keep a similar API to fluid-approximation simulations at high
+    levels of abstraction.
```

Comment:
We would also need a flexible enough `Simulation` (or something) class that incorporates mesh data (generically) and particle data (for PIC). Hell, if we wanted to have this be truly generic... some simulations are not going to have a mesh whatsoever. As an example: https://journals.aps.org/pre/abstract/10.1103/PhysRevE.98.033307

In other words, both mesh data and particle data might need to be optional *in the **very** long run*.